In [103]:
# initial setup
%run "../../../common/0_notebooks_base_setup.py"


/media/paulati/Nuevo vol/paula/dh/2021/dsad_2021_paula/common
default checking
Running command `conda list`... ok
jupyterlab=2.2.6 already installed
pandas=1.1.5 already installed
bokeh=2.2.3 already installed
seaborn=0.11.0 already installed
matplotlib=3.3.2 already installed
ipywidgets=7.5.1 already installed
pytest=6.2.1 already installed
chardet=4.0.0 already installed
psutil=5.7.2 already installed
scipy=1.5.2 already installed
statsmodels=0.12.1 already installed
scikit-learn=0.23.2 already installed
xlrd=2.0.1 already installed
Running command `conda install --yes nltk=3.5.0`... ok
Solving environment: ...working... done

# All requested packages already installed.


unidecode=1.1.1 already installed
pydotplus=2.0.2 already installed
pandas-datareader=0.9.0 already installed
flask=1.1.2 already installed


In [104]:
import pandas as pd
import numpy as np

<link rel="stylesheet" href="../../../common/dhds.css">
<div class="Table">
    <div class="Row">
        <div class="Cell grey left"> <img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_portada.png" align="center" /></div>
        <div class="Cell right">
            <div class="div-logo"><img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/common/logo_DH.png" align="center" width=70% /></div>
            <div class="div-curso">DATA SCIENCE</div>
            <div class="div-modulo">MÓDULO 2</div>
            <div class="div-contenido">Limpieza de Datos</div>
        </div>
    </div>
</div>

### Agenda

- Aprender operaciones involucradas en la limpieza de datos y los métodos asociados

- Aprender a usar herramientas para la limpieza de datos: apply, applymap y expresiones lambda

- Conocer e implementar las técnicas de "tidy data"

- Introducir el concepto de expresiones regulares y su utilidad en la limpieza de datos


<div class="div-dhds-fondo-1"> Limpieza y preparación de datos

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Limpieza y preparación de datos

---

En 2009 Mike Driscoll (data scientist y CEO de Metamarkets) popularizó el término **"data munging"** para referirse al **arduo proceso de limpiar, preparar y validar los datos** 


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_tiempo_data_scientist_1.png" align="center"></img>


Fuente: <a href="http://www.forbes.com/sites/gilpress/2016/03/23/data-preparation-most-time-consuming-least-enjoyable-data-science-task-survey-says/" target="_blank">http://www.forbes.com/sites/gilpress/2016/03/23/data-preparation-most-time-consuming-least-enjoyable-data-science-task-survey-says/</a>



### Limpieza y preparación de datos

---

**En 2013**, **Josh Wills** (ex director de Data Science de Cloudera y posterior Director of Data Engineering en Slack ) comenta:  **"I’m a data janitor. That’s the sexiest job of the 21st century. It’s very flattering, but it’s also a little baffling."**

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_data_preparation.png" align="center"></img>


Traducción: 
En Data Science, se invierte un 80% del tiempo en preparar los datos y el 20% restante en quejarse de la necesidad de preparar los datos


### Limpieza y preparación de datos

---

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_garbage_data.gif" align="center" width=450px></img>

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_garbage_data_acting_man.gif" align="center" width=450px></img>






### Limpieza y preparación de datos

---
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_misbehaving.jpg" align="center" width=50%></img>




<div class="div-dhds-fondo-1"> Limpieza de datos con 
Pandas


<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Limpieza de datos

---

La limpieza es un paso necesario en todo proyecto de datos. 
Podemos resumir el proceso de limpieza de datos refiriéndonos a las siguientes <u>cinco tareas</u>:

**1. Resolver problemas de formato y asignar los tipos correctos de datos**: 

Por ejemplo cuando al pasar de CSV a Pandas una fecha no se importa correctamente. Ej: 20090609231247 en lugar de 2009-06-09 23:12:47.

El formato en que se encuentran los datos va a afectar nuestro análisis por varias razones. Por ejemplo, las operaciones que se pueden realizar dependen del tipo de datos. Además algunos tipos ocupan menos espacio en memoria que otros.


**2. Estandarizar categorías**: 

Cuando los datos se recolectaron con un sistema que no tiene los valores tipificados, valores que representan las mismas categorías pueden estar expresados de forma distinta, por ejemplo Arg, AR y Argentina. 

**3. Corregir valores erróneos**: 

Por ejemplo un valor numérico o inválido para describir el género. O una edad representada por un número negativo o mucho mayor que 100.





**4. Completar datos faltantes**: 

Los datasets del mundo real suelen venir con datos faltantes que responden a información que se perdió o nunca se recolectó. 

Existen varias técnicas para completar datos faltantes. Al proceso de completar datos faltantes se lo llama **"imputación"**.

**5. Organizar correctamente el dataset**: 

Es importante estructurar las filas y columnas de la forma más conveniente. Para hacerlo se pueden aplicar las reglas del **"tidy data"**.


<div class="div-dhds-fondo-1"> Datos faltantes

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>

### Representación de datos faltantes

---

* Los datasets del mundo real siempre tienen **datos faltantes**

* Cada lenguaje/framework tiene su forma de lidiar con estos

* Pandas utiliza los valores **None**, **NaN** o **NaT** debido a que se basa en Numpy

    - **None**: objeto de Python que representa ausencia de dato
    
    - **NaN**: definición de valor faltante de floats 
    
    - **NaT**: se utiliza para valores faltantes del tipo Timestamp


### Valores faltantes

---

En general todo conjunto de datos suele tener datos faltantes (ya sea porque esos datos no fueron recolectado o nunca existieron)

* Debemos poder detectar, rellenar o eliminar datos faltantes

* Hay que utilizar conocimiento del dominio para definir cuáles datos faltantes se completarán y cómo. 

* Pandas ofrece varias formas de hacer esto.


### Valores faltantes

---

#### <u>Análisis con datos completos</u>.

Una forma de proceder consiste en eliminar los registros que presentan algún dato faltante:

* Ventaja: fácil implementación

* Desventaja: pérdida importante de información, posibles sesgos en los estimadores de los parámetros.

Este método asume que la falta de respuesta se generó de forma aleatoria, lo que por lo general no sucede. 

Si los datos que faltan son pocos y al azar, entonces este método es el ideal.


In [105]:
data_location = "../Data/melb_data.csv"

data = pd.read_csv(data_location)

print(data.shape)

data.head(3)



(13580, 21)


Suburb          Address  Rooms Type      Price Method SellerG  \
0  Abbotsford     85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  

[3 rows x 21 columns]

##### ```dropna```

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html

In [106]:
# elimino los registros que tienen algún campo con valor nulo

data.dropna(axis=0, how = 'any', inplace = True)

print(data.shape)

data.head(3)


(6196, 21)


Suburb          Address  Rooms Type      Price Method SellerG  \
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   
4  Abbotsford      55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
1        4019.0  
2        4019.0  
4        4019.0  

[3 rows x 21 columns]



#### <u>Métodos de imputación</u>

##### **Imputación Simple**:

Consiste en asignar un valor por cada dato faltante a partir de la propia variable o las demás variables para lograr una base de datos completa.


##### **Imputación Múltiple**:

A cada valor faltante se le asigna un grupo de <i>m</i> valores, generando <i>m</i> bases de datos completas. En cada una de las <i>m</i> bases de datos se estiman los parámetros de interés y luego se combinan los resultados obtenidos.



##### **Imputación por Media**:

Se puede completar los valores faltantes reemplazandolos **por la media** de la serie o **por la media condicionada** a determinada categoría. Por ejemplo, dado un valor de estatura faltante para una mujer, reemplazarlo por la media de las mujeres. 

Este enfoque tiene ventajas y desventajas:

* Ventaja: Es muy probable acercarme al verdadero valor del dato faltante

* Desventajas: 

    - Reduzco artificialmente la variabilidad y la aleatoriedad de los datos, lo cual me puede llevar a conclusiones equivocadas.

    - Si existía correlación entre esta variable y otras, ese valor puede verse afectado.






##### **Imputación Hot Deck**:

Consisten en completar con un dato existente de la muestra, siguiendo algún criterio. 

Pueden ser 

* aleatorios (se elige un elemento al azar), 
* secuencial (se completa con el valor inmediatamente anterior o posterior),
* vecino más cercano.

##### **Imputación por Regresión**:

Se emplean modelos de regresión para estimar el dato que falta a partir de las demás variables del dataset. 


<div class="div-dhds-fondo-1"> Introducción a imputación en Pandas

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Introducción a imputación en Pandas

---

El método <b>fillna()</b> de Pandas, permite varios tipos de imputación que puede especificarse en el parámetro “method”:

* Completar los datos con valores definidos por fuera del data set. (<b>method = None</b>) ya que se ingresa directamente como argumento el valor con el que se imputará. 

* Completar los datos faltantes con el valor anterior o el siguiente (ideal para series de tiempo) (<b>method = ffill</b>) o (<b>method = bfill</b>)

* Completar por la media, moda o la mediana. <b>dff.fillna(dff.mean())</b>


### Imputación Simple en Pandas

In [164]:
data_location = "../Data/melb_data.csv"

data = pd.read_csv(data_location)

print(data.shape)

data.BuildingArea.isnull().sum()


(13580, 21)


6450

Hay 6450 propiedades que tienen nulo en el dato BuildingArea. 

Vamos a completar los valores nulos en este campo asignando un valor al azar de ```data.BuildingArea```



In [108]:
data.BuildingArea.head(5)

0      NaN
1     79.0
2    150.0
3      NaN
4    142.0
Name: BuildingArea, dtype: float64

In [109]:
possible_values = data.BuildingArea[data.BuildingArea.notnull()]

print("cantidad de valores posibles: ", len(possible_values))

rng = np.random.default_rng()

# All values generated will be less than high:
fill_value_indexes = rng.integers(low = 0, high = len(possible_values), size = 1)

fill_value_index = fill_value_indexes[0]

print("índice del elemento que usamos para imputar: ", fill_value_index)

fill_value = possible_values.iloc[fill_value_index]

print("valor de imputación: ", fill_value)

cantidad de valores posibles:  7130
índice del elemento que usamos para imputar:  5862
valor de imputación:  116.65


In [110]:
data_BuildingArea_complete = data.BuildingArea.fillna(fill_value)
print("cantidad de nulos: ", data_BuildingArea_complete.isnull().sum())
data_BuildingArea_complete.head(5)

cantidad de nulos:  0


0    116.65
1     79.00
2    150.00
3    116.65
4    142.00
Name: BuildingArea, dtype: float64

### Imputación por Media en Pandas

Hay 6450 registros que tienen valor nulo en el campo BuildingArea

In [114]:
data.BuildingArea.isnull().sum()

6450

Vamos a completar estos registros con el valor medio del campo BuildingArea

In [116]:
data.BuildingArea.fillna(data.BuildingArea.mean(), inplace = True)

Y verificamos que no quedó ningún registro con valor nulo en BuildingArea

In [117]:
data.BuildingArea.isnull().sum()

0

### Imputación por Media Condicionada en Pandas

Vamos a calcular la media de BuildingArea por CouncilArea, y vamos a imputar los valores faltantes usando esta media condicionada

**Aviso**:
    
Los métodos presentados en esta parte se verán en profundidad en la clase de Pandas 2, pero los mencionamos aquí para presentar los pasos a seguir al realizar este tipo de imputación.

Calculamos la media de BuildingArea para cada valor de CouncilArea

In [169]:
fill_values = data.groupby('CouncilArea').BuildingArea.mean()

fill_values.head(5)

CouncilArea
Banyule       151.251855
Bayside       182.499739
Boroondara    179.629367
Brimbank      144.531272
Cardinia      251.646857
Name: BuildingArea, dtype: float64

In [124]:
type(fill_values)

pandas.core.series.Series

In [125]:
fill_values.index

Index(['Banyule', 'Bayside', 'Boroondara', 'Brimbank', 'Cardinia', 'Casey',
       'Darebin', 'Frankston', 'Glen Eira', 'Greater Dandenong', 'Hobsons Bay',
       'Hume', 'Kingston', 'Knox', 'Macedon Ranges', 'Manningham',
       'Maribyrnong', 'Maroondah', 'Melbourne', 'Melton', 'Monash',
       'Moonee Valley', 'Moorabool', 'Moreland', 'Nillumbik', 'Port Phillip',
       'Stonnington', 'Unavailable', 'Whitehorse', 'Whittlesea', 'Wyndham',
       'Yarra', 'Yarra Ranges'],
      dtype='object', name='CouncilArea')

In [126]:
fill_values.values

array([151.49627301, 172.0351117 , 170.59074421, 147.10187164,
       224.46161871, 156.21580652, 138.37248099, 154.25619429,
       149.21586009, 145.72075955, 150.18560848, 152.11110941,
       154.20511589, 171.78544245, 166.33361425, 194.35737229,
       134.14877026, 150.69470744, 121.56902506, 174.04584769,
       149.95013734, 150.21164629, 151.96764989, 136.84631178,
       162.81715828, 126.3945189 , 140.86686886, 151.96764989,
       154.5515432 , 162.35073738, 158.52770984, 131.53919485,
       150.08062217])

Construimos un DataFrame que con dos columnas (CouncilArea y BuildingArea) a partir de la instancia de Series calculada en el paso anterior

In [170]:
fill_values_df = pd.DataFrame(fill_values)
fill_values_df.reset_index(inplace=True)
fill_values_df.head(5)

CouncilArea  BuildingArea
0     Banyule    151.251855
1     Bayside    182.499739
2  Boroondara    179.629367
3    Brimbank    144.531272
4    Cardinia    251.646857

Combinamos el DataFrame original con los valores de las medias por CouncilArea

In [165]:
data = data.merge(fill_values_df, on = "CouncilArea", suffixes = ("", "_mean"), how = 'left')
data.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5    3067.0  ...  1.0     202.0           NaN        NaN   
1  4/02/2016       2.5    3067.0  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5    3067.0  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5    3067.0  ...  1.0      94.0           NaN        NaN   
4  4/06/2016       2.5    3067.0  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996   144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079   144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093   144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969   144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072   144.9941  Northern Metropolitan         4019.0   

  BuildingArea_mean  
0        131.539195  
1        131.539195  
2        131.539195  
3        131.539195  
4        131.539195  

[5 rows x 22 columns]

Usamos ```fillna``` para completar los valores de la columna ```BuildingArea``` usando los valores de la columna ```BuildingArea_mean```

In [167]:
data.BuildingArea.fillna(data.BuildingArea_mean, inplace = True)

data.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5    3067.0  ...  1.0     202.0    131.539195        NaN   
1  4/02/2016       2.5    3067.0  ...  0.0     156.0     79.000000     1900.0   
2  4/03/2017       2.5    3067.0  ...  0.0     134.0    150.000000     1900.0   
3  4/03/2017       2.5    3067.0  ...  1.0      94.0    131.539195        NaN   
4  4/06/2016       2.5    3067.0  ...  2.0     120.0    142.000000     2014.0   

   CouncilArea  Lattitude Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996   144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079   144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093   144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969   144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072   144.9941  Northern Metropolitan         4019.0   

  BuildingArea_mean  
0        131.539195  
1        131.539195  
2        131.539195  
3        131.539195  
4        131.539195  

[5 rows x 22 columns]

Vemos si quedan registros con valor nulo en el campo BuildingArea después de la imputación

In [172]:
data.BuildingArea.isnull().sum()

685

**¿Por qué quedaron algunos valores de BuildingArea en nulo?**

<div class="div-dhds-fondo-1"> Tidy Data

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Las reglas “tidy data”

---

Decimos que un dataset está ordenado cuando:

* Cada variable es una columna

* Cada observación es una fila

* Cada tipo de unidad observacional forma una tabla

Algunas definiciones:

* Variable: Es la medición de un atributo, por ejemplo, peso, altura, etc 

* Valor: Es la medida que toma una variable para una observación

* Observación: Todas las observaciones toman el mismo tipo de valores para cada variable. 




### Tidy data. Ejemplo:

---

<table><row>    
<td style="border-right: 1px solid black;">
<b>Messy data</b><br/>        
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_messy_data.png" align="center"/>
</td><td> </td>
<td><b>Tidy data</b><br/>
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_tidy_data.png" align="center" />
</td></row></table>

* Cada unidad observacional debería ser un tratamiento sobre una persona

* En los formatos “messy data”, si quisiera agregar tratamientos que no aplican a todas las personas se llenaría el dataset de valores nulos.


<div class="div-dhds-fondo-1"> Herramientas para la limpieza de datos

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>

### Expresiones lambda

---

Recordemos cómo construir una función usando ```def:```

Este forma de declarar funciones sirve, entre otras cosas, para no duplicar y para modularizar el código.



In [112]:
def square(x):
    return x ** 2

square(10)

100

¿Tendría sentido definir una función de este modo si vamos a invocar dicha función una única vez?

Las funciones ```lambda``` pueden ser directamente definidas en el código que las va a utilizar y sin necesidad de otorgarles un nombre (son **funciones anónimas**).

**Toman una única expresión y no contienen la expresión ```return```**.

Se definen en una única línea.


In [113]:
square = lambda x: x ** 2

square(10)

100

### ```apply()``` en Pandas

---

- El método ```apply()``` permite aplicar cualquier función a los elementos de un DataFrame. 

    Se puede utilizar:

    * Por columna: ```df.apply(mi_funcion)``` (iterando por filas)
    
    * Por fila  ```df.apply(mi_funcion, axis = 1)``` (iterando por columnas)

    * Elemento por elemento ```df.applymap(mi_funcion)```


- ```.apply()``` también se puede utilizar sobre una instancia de Series, elemento por elemento.


- Una buena propiedad de ```apply()``` es que permite aplicar las operaciones vectorizadas de NumPy.




#### Ejemplo ```apply()``` y ```lambda```

In [173]:
data_location = "../Data/melb_data.csv"

data = pd.read_csv(data_location)

print(data.shape)

data.head(3)


(13580, 21)


Suburb          Address  Rooms Type      Price Method SellerG  \
0  Abbotsford     85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  

[3 rows x 21 columns]

Vamos a calcular el precio por m2 teniendo en cuenta que no podemos dividir por 0:

In [175]:
precio_m2 = data.apply(lambda x: 0 if x['Landsize'] == 0 else x['Price'] / x['Landsize'], axis=1)

print(type(precio_m2))


<class 'pandas.core.series.Series'>


In [177]:
precio_m2.head(5)

0     7326.732673
1     6634.615385
2    10932.835821
3     9042.553191
4    13333.333333
dtype: float64

<div class="div-dhds-fondo-1"> Expresiones regulares

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Expresiones regulares

---

¿Qué son las expresiones regulares y para qué sirven?


Una expresión regular es una **secuencia de caracteres** que define un **patrón de búsqueda** de texto. Las regex constituyen un lenguaje muy flexible que sirve para identificar y extraer información de un cuerpo de caracteres **no estructurado**.

Ejemplos:

* identificar direcciones de correo electrónico

* identificar DNI, CUIT, CUIL, etc.




### Metacaracteres especiales

---

Los **metacaracteres especiales** son aquellos que indican que se debe matchear algo de forma **no literal**, o que afectan a otras partes de la regex repitiendo caracteres o cambiando su significado.



<u>Algunos metacaracteres frecuentes</u>:

**\d** Cualquier dígito del 0 al 9

**\w** Cualquier caracter alfanumérico (A-Z, a-z, 0-9 y _)

**\s** Cualquier espacio en blanco (espacio, tabulado, nueva línea, etc.)

**.** Cualquier caracter con excepción de nueva línea (\n)


Y sus complementos:

**\D** Todo menos cualquier dígito

**\W** Todo menos cualquier carácter alfanumérico

**\S** Todo menos cualquier espacio en blanco

<u>Cuantificadores</u>:

**\*** Cero o más del elemento anterior

**+** Uno o más del elemento anterior

**{min,max}** Definen mínimos y máximos de repetición

**?** Opcional, ó cero o uno del elemento anterior



<u>Ejemplos</u>:

**\d+** Encuentra un número o más

**.\*** Encuentra cero o más de cualquier caracter

**\w{2,6}** Encuentra un conjunto de caracteres alfanuméricos con una longitud que va de 2 a 6 caracteres

<u>Otros metacaracteres importantes</u>:

**[ ]** Definen un conjunto de caracteres

**[A-Za-z0-9_]** Definen rangos

**( )** Definen un grupo de captura

**|** Operador "o"

**\** Escapa los caracteres especiales

Metacaracteres de posición:

**^** Comienzo de un string (o negación dentro de un conjunto, [^ ])

**$** Final de un string

**\b** Límite de palabra



### ¿Qué patrón pueden identificar para los siguientes strings?

---

ana_laura@hotmail.com

juan.perez@gmail.com

florgimenez@yahoo.com.ar


Queremos una expresión que sirva como patrón para cualquiera de las tres direcciones de email presentadas.

Cada par de paréntesis ( ) define un grupo 

In [198]:
import re

patron = "^(?P<nombre_de_usuario>.*)@(?P<dominio>.*)(\.com)(\.ar)?$"

regex = re.compile(patron, flags = re.IGNORECASE)

texto1 = "ana_laura@hotmail.com"

regex.search(texto1).groups()

('ana_laura', 'hotmail', '.com', None)

In [199]:
regex.search(texto1).group(0)


'ana_laura@hotmail.com'

In [200]:
regex.search(texto1).group(1)

'ana_laura'

In [201]:
regex.search(texto1).group(2)

'hotmail'

In [202]:
regex.search(texto1).group(3)

'.com'

In [207]:
regex.search(texto1).group(4)

In [197]:
regex.search(texto1).group("nombre_de_usuario")

'ana_laura'

In [208]:
regex.search(texto1).group("dominio")

'hotmail'

In [210]:
texto2 = "juan.perez@gmail.com"

regex.search(texto2).groups()


('juan.perez', 'gmail', '.com', None)

In [211]:
texto3 = "florgimenez@yahoo.com.ar"

regex.search(texto3).groups()


('florgimenez', 'yahoo', '.com', '.ar')

In [212]:
regex.search(texto3).group(4)

'.ar'

<div class="div-dhds-fondo-1"> Hands-on
<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


### Ejercicio 1

Usando los datos de propiedades de Melbourne (../Data/melb_data.csv), queremos imputar los valores faltantes del campo ```YearBuilt``` usando la mediana de ese campo por vendedor ```SellerG```

¿Cántos valores nulos en el campo ```YearBuilt``` hay antes de la imputación?

¿Cántos valores nulos en el campo ```YearBuilt``` hay después de la imputación?




### Ejercicio 2


Queremos definir una expresión regular que nos permita extraer el punto cardinal al que referencia Regionname en los datos de propiedades de Melbourne (../Data/melb_data.csv), 

Por ejemplo:

* si el valor es Northern Metropolitan o Northern Victoria, el resultado debe ser North

* si el valor es Western Metropolitan o Western Victoria, el resultado debe ser West

* si el valor es Southern Metropolitan o Southern Victoria, el resultado debe ser South

* si el valor es Eastern Metropolitan o Eastern Victoria, el resultado debe ser East






Sugerencia: usar como textos para hacer pruebas 

* 'Northern Metropolitan', 

* 'Western Metropolitan', 

* 'Southern Metropolitan', 

* 'Eastern Metropolitan', 

* 'South-Eastern Metropolitan', 

* 'Eastern Victoria',

* 'Northern Victoria', 

* 'Western Victoria' 

y el método ```search```

### Ejercicio 3

Usando la expresión regular definida en el punto anterior, el método ```apply``` y una función ```lambda```, obtener una instancia de Series con los puntos cardinales a los que referencia ```Regionname``` en cada registro.

### Solución

---


### Ejercicio 1


In [219]:
data_location = "../Data/melb_data.csv"

data = pd.read_csv(data_location)

print("cantidad de nulos antes de la imputación: ", data.YearBuilt.isnull().sum())


cantidad de nulos antes de la imputación:  5375


In [220]:
fill_values = data.groupby('SellerG').YearBuilt.median()

fill_values.head(5)

SellerG
@Realty         1970.0
ASL             1962.5
Abercromby's    1946.5
Ace                NaN
Alexkarbon      2000.0
Name: YearBuilt, dtype: float64

In [221]:
fill_values_df = pd.DataFrame(fill_values)
fill_values_df.reset_index(inplace=True)
fill_values_df.head(5)

SellerG  YearBuilt
0       @Realty     1970.0
1           ASL     1962.5
2  Abercromby's     1946.5
3           Ace        NaN
4    Alexkarbon     2000.0

In [222]:
data = data.merge(fill_values_df, on = "SellerG", suffixes = ("", "_median"), how = 'left')
data.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5    3067.0  ...  1.0     202.0           NaN        NaN   
1  4/02/2016       2.5    3067.0  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5    3067.0  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5    3067.0  ...  1.0      94.0           NaN        NaN   
4  4/06/2016       2.5    3067.0  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996   144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079   144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093   144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969   144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072   144.9941  Northern Metropolitan         4019.0   

  YearBuilt_median  
0           1970.0  
1           1970.0  
2           1970.0  
3           1970.0  
4           1960.0  

[5 rows x 22 columns]

In [223]:
data.YearBuilt.fillna(data.YearBuilt_median, inplace = True)

data.head(5)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5    3067.0  ...  1.0     202.0           NaN     1970.0   
1  4/02/2016       2.5    3067.0  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5    3067.0  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5    3067.0  ...  1.0      94.0           NaN     1970.0   
4  4/06/2016       2.5    3067.0  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996   144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079   144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093   144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969   144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072   144.9941  Northern Metropolitan         4019.0   

  YearBuilt_median  
0           1970.0  
1           1970.0  
2           1970.0  
3           1970.0  
4           1960.0  

[5 rows x 22 columns]

In [224]:
print("cantidad de nulos después de la imputación: ", data.YearBuilt.isnull().sum())

cantidad de nulos después de la imputación:  51


**¿A qué vendedores corrresponden los 51 registros que quedaron en nulo?**

### Ejercicio 2


In [241]:
data_location = "../Data/melb_data.csv"

data = pd.read_csv(data_location)

In [240]:
patron = "^(?P<punto_cardinal>.*)ern\s(Metropolitan|Victoria)$"

regex = re.compile(patron, flags = re.IGNORECASE)


In [230]:
texto1 = 'Northern Metropolitan'

regex.search(texto1).group("punto_cardinal")

'North'

In [231]:
texto2 = 'Western Metropolitan'

regex.search(texto2).group("punto_cardinal")

'West'

In [232]:
texto3 = 'Southern Metropolitan'

regex.search(texto3).group("punto_cardinal")

'South'

In [234]:
texto4 = 'Southern Metropolitan'

regex.search(texto4).group("punto_cardinal")

'South'

In [235]:
texto5 = "South-Eastern Metropolitan"

regex.search(texto5).group("punto_cardinal")

'South-East'

In [236]:
texto6 = "Eastern Victoria"

regex.search(texto6).group("punto_cardinal")

'East'

In [237]:
texto7 = "Northern Victoria"

regex.search(texto7).group("punto_cardinal")

'North'

In [238]:
texto8 = "Western Victoria"

regex.search(texto8).group("punto_cardinal")

'West'

### Ejercicio 3

In [242]:
patron = "^(?P<punto_cardinal>.*)ern\s(Metropolitan|Victoria)$"

regex = re.compile(patron, flags = re.IGNORECASE)


In [253]:
resultado = data.Regionname.apply(lambda x: regex.search(x).group("punto_cardinal"))

resultado.sample(10)

1847          South
791           South
12593         North
8928          North
8126           West
2457           West
11446          East
120           North
11335    South-East
5220          North
Name: Regionname, dtype: object

In [254]:
type(resultado)

pandas.core.series.Series

<div class="div-dhds-fondo-1"> Referencias

<img src="https://raw.githubusercontent.com/Digital-House-DATA/ds_blend_2021_img/master/M2/CLASE_09_LimpiezaDeDatos/Presentacion/img/M2_CLASE_09_separador.png" align="center"></img>    
</div>


Tutorial de expresiones regulares:

<a href="https://github.com/Digital-House-DATA/ds_blend_2021_img/wiki/Tutorial-de-Expresiones-Regulares-(regex)" target="_blank">Tutorial de Expresiones Regulares (regex)</a>

Documentación:

<a href="https://docs.python.org/3/library/re.html" target="_blank">re</a>

<a href="https://docs.python.org/3/howto/regex.html" target="_blank">regex</a>


Sitios web para practicar:

<a href="https://regexr.com/" target="_blank">regexr</a>

<a href="https://regex101.com/" target="_blank">regex101</a>

Missing Data

https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html

https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html#filling-missing-values-fillna